File: chatbot/packages/mobile-sdk-android/src/main/java/com/wealtharena/mobile/sdk/WealthArenaClient.kt
package com.wealtharena.mobile.sdk

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeout
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.*
import java.util.concurrent.TimeUnit

/**
 * WealthArena Mobile SDK for Android
 * Main client class for Android integration
 */

// Data classes
data class ChatRequest(
    val message: String,
    val symbol: String? = null,
    val mode: String? = null
)

data class ChatResponse(
    val reply: String,
    val sources: List<String>,
    val suggestions: List<String>,
    val timestamp: String
)

data class AnalyzeRequest(
    val symbol: String
)

data class AnalyzeResponse(
    val symbol: String,
    val current_price: Double,
    val indicators: Indicators,
    val signals: List<Signal>,
    val timestamp: String
)

data class Indicators(
    val sma_20: List<Double?>,
    val ema_12: List<Double?>,
    val rsi: List<Double?>
)

data class Signal(
    val type: String,
    val indicator: String,
    val message: String,
    val explanation: String
)

data class TradeRequest(
    val action: String,
    val symbol: String,
    val quantity: Double,
    val price: Double? = null
)

data class TradeResponse(
    val success: Boolean,
    val message: String,
    val new_balance: Double,
    val position: Position? = null,
    val timestamp: String
)

data class Position(
    val quantity: Double,
    val avg_price: Double
)

data class StateResponse(
    val balance: Double,
    val positions: Map<String, Position>,
    val trades: List<Trade>,
    val total_pnl: Double,
    val timestamp: String
)

data class Trade(
    val timestamp: String,
    val symbol: String,
    val action: String,
    val quantity: Double,
    val price: Double,
    val total: Double
)

data class LearnResponse(
    val lessons: List<Lesson>,
    val quizzes: List<Quiz>,
    val timestamp: String
)

data class Lesson(
    val id: String,
    val title: String,
    val content: String,
    val topics: List<String>
)

data class Quiz(
    val id: String,
    val question: String,
    val options: List<String>,
    val correct: Int,
    val explanation: String
)

data class HealthResponse(
    val status: String,
    val service: String,
    val version: String,
    val timestamp: String
)

// API Interface
interface WealthArenaApi {
    @POST("/v1/chat")
    suspend fun chat(@Body request: ChatRequest): ChatResponse

    @POST("/v1/analyze")
    suspend fun analyze(@Body request: AnalyzeRequest): AnalyzeResponse

    @GET("/v1/state")
    suspend fun state(): StateResponse

    @POST("/v1/papertrade")
    suspend fun paperTrade(@Body request: TradeRequest): TradeResponse

    @GET("/v1/learn")
    suspend fun learn(): LearnResponse

    @GET("/v1/healthz")
    suspend fun health(): HealthResponse
}

// Custom Exception
class WealthArenaException(
    message: String,
    val code: String? = null,
    val status: Int? = null
) : Exception(message)

// Main Client Class
class WealthArenaClient private constructor(
    private val api: WealthArenaApi,
    private val timeoutMs: Long = 30000,
    private val maxRetries: Int = 3
) {

    data class Builder(
        private var baseUrl: String,
        private var token: String? = null,
        private var timeoutMs: Long = 30000,
        private var maxRetries: Int = 3,
        private var enableLogging: Boolean = false
    ) {
        fun token(token: String?) = apply { this.token = token }
        fun timeout(timeoutMs: Long) = apply { this.timeoutMs = timeoutMs }
        fun retries(maxRetries: Int) = apply { this.maxRetries = maxRetries }
        fun enableLogging(enable: Boolean) = apply { this.enableLogging = enable }

        fun build(): WealthArenaClient {
            val httpClient = OkHttpClient.Builder()
                .connectTimeout(timeoutMs, TimeUnit.MILLISECONDS)
                .readTimeout(timeoutMs, TimeUnit.MILLISECONDS)
                .writeTimeout(timeoutMs, TimeUnit.MILLISECONDS)
                .apply {
                    if (enableLogging) {
                        val logging = HttpLoggingInterceptor()
                        logging.level = HttpLoggingInterceptor.Level.BODY
                        addInterceptor(logging)
                    }
                    
                    if (token != null) {
                        addInterceptor { chain ->
                            val request = chain.request().newBuilder()
                                .addHeader("Authorization", "Bearer $token")
                                .build()
                            chain.proceed(request)
                        }
                    }
                }
                .build()

            val retrofit = Retrofit.Builder()
                .baseUrl(baseUrl)
                .client(httpClient)
                .addConverterFactory(GsonConverterFactory.create())
                .build()

            val api = retrofit.create(WealthArenaApi::class.java)
            return WealthArenaClient(api, timeoutMs, maxRetries)
        }
    }

    companion object {
        fun builder(baseUrl: String): Builder = Builder(baseUrl)
    }

    private suspend fun <T> executeWithRetry(operation: suspend () -> T): T {
        var lastException: Exception? = null
        
        repeat(maxRetries) { attempt ->
            try {
                return withTimeout(timeoutMs) {
                    withContext(Dispatchers.IO) {
                        operation()
                    }
                }
            } catch (e: Exception) {
                lastException = e
                if (attempt < maxRetries - 1) {
                    // Exponential backoff
                    val delay = (2.0.pow(attempt) * 1000).toLong()
                    kotlinx.coroutines.delay(delay)
                }
            }
        }
        
        throw lastException ?: WealthArenaException("Request failed after $maxRetries retries")
    }

    /**
     * Chat with the educational trading bot
     */
    suspend fun chat(request: ChatRequest): ChatResponse {
        return executeWithRetry { api.chat(request) }
    }

    /**
     * Analyze an asset with technical indicators
     */
    suspend fun analyze(request: AnalyzeRequest): AnalyzeResponse {
        return executeWithRetry { api.analyze(request) }
    }

    /**
     * Get current trading state
     */
    suspend fun state(): StateResponse {
        return executeWithRetry { api.state() }
    }

    /**
     * Execute a paper trade
     */
    suspend fun paperTrade(request: TradeRequest): TradeResponse {
        return executeWithRetry { api.paperTrade(request) }
    }

    /**
     * Get educational content and quizzes
     */
    suspend fun learn(): LearnResponse {
        return executeWithRetry { api.learn() }
    }

    /**
     * Health check
     */
    suspend fun health(): HealthResponse {
        return executeWithRetry { api.health() }
    }
}

// Extension function for easy client creation
fun createWealthArenaClient(baseUrl: String, token: String? = null): WealthArenaClient {
    return WealthArenaClient.builder(baseUrl)
        .token(token)
        .build()
}

File: chatbot/packages/mobile-sdk-ios/Sources/WealthArenaSDK/WealthArenaClient.swift
import Foundation

// MARK: - Configuration

public struct WealthArenaConfig {
    public let baseURL: URL
    public let token: String?
    public let timeout: TimeInterval
    public let maxRetries: Int
    
    public init(
        baseURL: URL,
        token: String? = nil,
        timeout: TimeInterval = 30.0,
        maxRetries: Int = 3
    ) {
        self.baseURL = baseURL
        self.token = token
        self.timeout = timeout
        self.maxRetries = maxRetries
    }
}

// MARK: - Main Client

@available(iOS 13.0, macOS 10.15, *)
public class WealthArenaClient {
    private let config: WealthArenaConfig
    private let session: URLSession
    
    public init(config: WealthArenaConfig) {
        self.config = config
        
        let sessionConfig = URLSessionConfiguration.default
        sessionConfig.timeoutIntervalForRequest = config.timeout
        sessionConfig.timeoutIntervalForResource = config.timeout
        
        self.session = URLSession(configuration: sessionConfig)
    }
    
    public convenience init(baseURL: URL, token: String? = nil) {
        let config = WealthArenaConfig(baseURL: baseURL, token: token)
        self.init(config: config)
    }
    
    // MARK: - Private Methods
    
    private func makeRequest<T: Codable>(
        endpoint: String,
        method: HTTPMethod = .GET,
        body: Codable? = nil,
        responseType: T.Type
    ) async throws -> T {
        guard let url = URL(string: "\(config.baseURL.absoluteString)/v1\(endpoint)") else {
            throw WealthArenaError.invalidURL
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = method.rawValue
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        if let token = config.token {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        
        if let body = body {
            do {
                request.httpBody = try JSONEncoder().encode(body)
            } catch {
                throw WealthArenaError.decodingError(error)
            }
        }
        
        return try await executeWithRetry(request: request, responseType: responseType)
    }
    
    private func executeWithRetry<T: Codable>(
        request: URLRequest,
        responseType: T.Type
    ) async throws -> T {
        var lastError: Error?
        
        for attempt in 0..<config.maxRetries {
            do {
                let (data, response) = try await session.data(for: request)
                
                guard let httpResponse = response as? HTTPURLResponse else {
                    throw WealthArenaError.noData
                }
                
                guard httpResponse.statusCode == 200 else {
                    let errorMessage = String(data: data, encoding: .utf8) ?? "Unknown error"
                    throw WealthArenaError.serverError(httpResponse.statusCode, errorMessage)
                }
                
                do {
                    let result = try JSONDecoder().decode(responseType, from: data)
                    return result
                } catch {
                    throw WealthArenaError.decodingError(error)
                }
                
            } catch {
                lastError = error
                
                if attempt < config.maxRetries - 1 {
                    // Exponential backoff
                    let delay = pow(2.0, Double(attempt)) * 1.0
                    try await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
                }
            }
        }
        
        throw lastError ?? WealthArenaError.unknown(NSError(domain: "WealthArena", code: -1, userInfo: nil))
    }
    
    // MARK: - Public API Methods
    
    /// Chat with the educational trading bot
    public func chat(_ request: ChatRequest) async throws -> ChatResponse {
        return try await makeRequest(
            endpoint: "/chat",
            method: .POST,
            body: request,
            responseType: ChatResponse.self
        )
    }
    
    /// Analyze an asset with technical indicators
    public func analyze(_ request: AnalyzeRequest) async throws -> AnalyzeResponse {
        return try await makeRequest(
            endpoint: "/analyze",
            method: .POST,
            body: request,
            responseType: AnalyzeResponse.self
        )
    }
    
    /// Get current trading state
    public func state() async throws -> StateResponse {
        return try await makeRequest(
            endpoint: "/state",
            method: .GET,
            responseType: StateResponse.self
        )
    }
    
    /// Execute a paper trade
    public func paperTrade(_ request: TradeRequest) async throws -> TradeResponse {
        return try await makeRequest(
            endpoint: "/papertrade",
            method: .POST,
            body: request,
            responseType: TradeResponse.self
        )
    }
    
    /// Get educational content and quizzes
    public func learn() async throws -> LearnResponse {
        return try await makeRequest(
            endpoint: "/learn",
            method: .GET,
            responseType: LearnResponse.self
        )
    }
    
    /// Health check
    public func health() async throws -> HealthResponse {
        return try await makeRequest(
            endpoint: "/healthz",
            method: .GET,
            responseType: HealthResponse.self
        )
    }
}

// MARK: - HTTP Method

private enum HTTPMethod: String {
    case GET = "GET"
    case POST = "POST"
    case PUT = "PUT"
    case DELETE = "DELETE"
}

// MARK: - Convenience Factory

public func createWealthArenaClient(baseURL: URL, token: String? = nil) -> WealthArenaClient {
    return WealthArenaClient(baseURL: baseURL, token: token)
}

